File: LeetCode/ArrayStrings/java/removeDuplicates.java
package LeetCode.ArrayStrings.java;

import java.util.Arrays;

record removeDuplicatesRecord(int[] nums, int[] expected) {
}

public class removeDuplicates {
    public int removeDuplicatesSoln(int[] nums) {
        int l = 0;
        int r = 0;
        int n = nums.length;

        while (r < n) {
            int count = 1;

            while ((r + 1 < n) && (nums[r] == nums[r + 1])) {
                r += 1;
                count += 1;
            }

            for (int i = 0; i < Math.min(1, count); i += 1) {
                nums[l] = nums[r];
                l += 1;
            }

            r += 1;
        }

        return l;
    }

    public static void main(String[] args) {
        removeDuplicates soln = new removeDuplicates();
        removeDuplicatesRecord input1 = new removeDuplicatesRecord(
                new int[] { 1, 1, 2 },
                new int[] { 1, 2 });

        System.out.println("Input: nums: " + Arrays.toString(input1.nums()));
        System.out.println("Expected: " + Arrays.toString(input1.expected()));
        int k1 = soln.removeDuplicatesSoln(input1.nums());
        System.out.println("Result: " + Arrays.toString(Arrays.stream(input1.nums()).limit(k1).toArray()));
        System.out.println("-".repeat(50));

        removeDuplicatesRecord input2 = new removeDuplicatesRecord(
                new int[] { 0, 0, 1, 1, 1, 2, 2, 3, 3, 4 },
                new int[] { 0, 1, 2, 3, 4 });
        System.out.println("Input: nums: " + Arrays.toString(input2.nums()));
        System.out.println("Expected: " + Arrays.toString(input2.expected()));
        int k2 = soln.removeDuplicatesSoln(input2.nums());
        System.out.println("Result: " + Arrays.toString(Arrays.stream(input2.nums()).limit(k2).toArray()));
        System.out.println("-".repeat(50));
    }
}

File: LeetCode/ArrayStrings/python/removeDuplicates.py
from typing import List


class Solution:
    def removeDuplicates(self, nums: List[int]) -> int:
        l = 0
        r = 0
        n = len(nums)

        while r < n:
            count = 1

            while (r + 1 < n) and (nums[r] == nums[r + 1]):
                r += 1 
                count += 1

            for i in range(min(1, count)):
                nums[l] = nums[r]
                l += 1
            
            r += 1

        return l

if __name__ == "__main__":
    obj = Solution()

    input = {"nums" : [1,1,2], "expected" : [1,2]}
    print("Input: nums: {}".format(input["nums"]))
    print("Expected: {}".format(input["expected"]))
    k = obj.removeDuplicates(input["nums"])
    print("Result: {}".format(input["nums"][0:k]))
    print("-" * 50)

    input = {"nums" : [0,0,1,1,1,2,2,3,3,4], "expected" : [0,1,2,3,4]}
    print("Input: nums: {}".format(input["nums"]))
    print("Expected: {}".format(input["expected"]))
    k = obj.removeDuplicates(input["nums"])
    print("Result: {}".format(input["nums"][0:k]))
    print("-" * 50)